hail/python/hailtop/aiotools/stream.py

```diff
@@ -99,6 +106,20 @@ async def read(self, n: int = -1) -> bytes:
             return await blocking_to_async(self._thread_pool, self._f.read)
         return await blocking_to_async(self._thread_pool, self._f.read, n)
 
+    def _readexactly(self, n: int) -> bytes:
+        assert n >= 0
+        data: List[bytes] = []
+        while n > 0:
+            block = self._f.read(n)
+            if len(block) != n:
+                raise IncompleteRead()
```

Comment:
I think this should be if the `len(block) == 0`. [`RawIOBase.read`](https://docs.python.org/3/library/io.html#io.RawIOBase.read) is allowed to return fewer than `n` bytes. The docs indicate that a zero length return value when n is non-zero indicates EOF:
> If 0 bytes are returned, and size was not 0, this indicates end of file.

Reply:
Also, I think we should use UnexpectedEOFError from fs.py. It's already used for exactly this kind of thing in the copy tool.

Reply:
It's not used in the copy tool. I moved it to stream.py anyways because of import issues.